### PR TITLE
GH-34206: [C++] Don't let jemalloc defines affect unity builds

### LINKED ISSUE #34206: [C++] jemalloc prepocessor definitions causes unity builds to mix jemalloc and system allocator functions
Since `jemalloc.h` does

    #  define malloc je_malloc

when `JEMALLOC_MANGLE` is defined, we can get this error in CI during an unity build

    /arrow/cpp/src/arrow/vendored/ProducerConsumerQueue.h: In constructor 'arrow_vendored::folly::ProducerConsumerQueue<T>::ProducerConsumerQueue(uint32_t)':

    /arrow/cpp/src/arrow/vendored/ProducerConsumerQueue.h:82:39: error: 'je_arrow_malloc' is not a member of 'std'; did you mean 'je_arrow_malloc'?

       82 |         records_(static_cast<T*>(std::malloc(sizeof(T) * size))),

          |                                       ^~~~~~

    jemalloc_ep-prefix/src/jemalloc_ep/dist/include/jemalloc/jemalloc.h:254:32: note: 'je_arrow_malloc' declared here

      254 |     void JEMALLOC_SYS_NOTHROW *je_malloc(size_t size)

          |                                ^~~~~~~~~

    /arrow/cpp/src/arrow/vendored/ProducerConsumerQueue.h: In destructor 'arrow_vendored::folly::ProducerConsumerQueue<T>::~ProducerConsumerQueue()':

    /arrow/cpp/src/arrow/vendored/ProducerConsumerQueue.h:106:10: error: 'je_arrow_free' is not a member of 'std'; did you mean 'je_arrow_free'?

      106 |     std::free(records_);

          |          ^~~~

    jemalloc_ep-prefix/src/jemalloc_ep/dist/include/jemalloc/jemalloc.h:269:43: note: 'je_arrow_free' declared here

      269 | JEMALLOC_EXPORT void JEMALLOC_SYS_NOTHROW je_free(void *ptr)

          |                                           ^~~~~~~

### Component(s)

C++

### PR DESCRIPTION
**Pull request by felipecrv**
Since `jemalloc.h` does

    #  define malloc je_malloc

when `JEMALLOC_MANGLE` is defined, we can get this error in CI during an
unity build

    /arrow/cpp/src/arrow/vendored/ProducerConsumerQueue.h: In constructor 'arrow_vendored::folly::ProducerConsumerQueue<T>::ProducerConsumerQueue(uint32_t)':

    /arrow/cpp/src/arrow/vendored/ProducerConsumerQueue.h:82:39: error: 'je_arrow_malloc' is not a member of 'std'; did you mean 'je_arrow_malloc'?

       82 |         records_(static_cast<T*>(std::malloc(sizeof(T) * size))),

          |                                       ^~~~~~

    jemalloc_ep-prefix/src/jemalloc_ep/dist/include/jemalloc/jemalloc.h:254:32: note: 'je_arrow_malloc' declared here

      254 |     void JEMALLOC_SYS_NOTHROW *je_malloc(size_t size)

          |                                ^~~~~~~~~

    /arrow/cpp/src/arrow/vendored/ProducerConsumerQueue.h: In destructor 'arrow_vendored::folly::ProducerConsumerQueue<T>::~ProducerConsumerQueue()':

    /arrow/cpp/src/arrow/vendored/ProducerConsumerQueue.h:106:10: error: 'je_arrow_free' is not a member of 'std'; did you mean 'je_arrow_free'?

      106 |     std::free(records_);

          |          ^~~~

    jemalloc_ep-prefix/src/jemalloc_ep/dist/include/jemalloc/jemalloc.h:269:43: note: 'je_arrow_free' declared here

      269 | JEMALLOC_EXPORT void JEMALLOC_SYS_NOTHROW je_free(void *ptr)

          |                                           ^~~~~~~

* Closes: #34206

### CODE REVIEW

**Comment by felipecrv** [comment]
@zeroshade @assignUser

**Comment by assignUser** [comment]
Mhm this is out of my wheelhouse so I'll defer to someone else :D

**Comment by felipecrv** [comment]
I will extract the second commit as a separate PR if it fixes the MSVC linker errors.

**Comment by lidavidm** [comment]
Could also figure out how/why the jemalloc header is getting included there in the first place - I don't think we intended jemalloc to replace libc malloc everywhere

**Comment by felipecrv** [comment]
> Could also figure out how/why the jemalloc header is getting included there in the first place - I don't think we intended jemalloc to replace libc malloc everywhere

Is it even possible to not include that header in the compilation unit of a unity build?

**Comment by lidavidm** [comment]
Ah...you could `#ifdef ... #undef` (I guess you'd have to redefine it though), or possibly, we should consider whether we want jemalloc to `#define malloc` in the first place (I don't think that's what we intended)

**Comment by lidavidm** [comment]
If this unsticks the build though maybe let's go with this and file an issue for followup (on what we want to do with jemalloc)

**Comment by felipecrv** [comment]
Thank you for the suggestion: I will create an issue.

I think the defining of malloc and free can be disabled if we don't define `JEMALLOC_MANGLE` like it's done here

https://github.com/apache/arrow/blob/798b417ec342051f3ff30b9953dcb1f1b71dc9e8/cpp/src/arrow/memory_pool_jemalloc.cc#L28-L34

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (felipecrv) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2295]: My previous attempt at this change mistakenly deleted a crucial line. :sweat:

Fingers crossed that this will fix the MSVC issue.

**Comment by lidavidm** [comment]
We do want the mangle so that our copy of jemalloc does not conflict with any copies of jemalloc used by applications - so it seems unfortunate that mangling _also_ makes jemalloc try to redefine malloc. Possibly we can insert an `#undef malloc` after including jemalloc.

**Comment by zeroshade** [comment]
I second @lidavidm's suggestion of inserting an `#undef malloc` after including jemalloc

**Comment by felipecrv** [comment]
@lidavidm thank you for clarifying the intentions of the jemalloc integration. I didn't have enough context about this. I will wait for the current round of builds to finish and send an alternative fix that doesn't remove the `std::`

**Comment by felipecrv** [comment]
@lidavidm @zeroshade please take a look now.

After this is merged I will rebase my REE PR and hopefully get it to pass CI.

**Comment by kou** [comment]
I think that we should not mix `memory_pool_jemalloc.cc` to other sources instead of `#undef`:

```diff
diff --git a/cpp/src/arrow/CMakeLists.txt b/cpp/src/arrow/CMakeLists.txt
index 6382fdfb3a..33cf3d396d 100644
--- a/cpp/src/arrow/CMakeLists.txt
+++ b/cpp/src/arrow/CMakeLists.txt
@@ -243,6 +243,8 @@ set(ARROW_SRCS
 
 if(ARROW_JEMALLOC)
   list(APPEND ARROW_SRCS memory_pool_jemalloc.cc)
+  set_source_files_properties(memory_pool_jemalloc.cc
+                              PROPERTIES SKIP_UNITY_BUILD_INCLUSION ON)
 endif()
 
 append_avx2_src(util/bpacking_avx2.cc)
```

**Comment by felipecrv** [comment]
@kou yeah, that makes total sense. Done.

**Comment by kou** [comment]
Did you confirm that the change fix the problem on your environment?

**Comment by felipecrv** [comment]
@kou the issue manifested in https://github.com/apache/arrow/pull/33641 which is currently going through the CI build with this PR's commit on top. A true confirmation is a green build there.

**Review by zeroshade — APPROVED** [review]
LGTM

**Review by kou — APPROVED** [review]
+1

**Pull request merged**
